### PR TITLE
Remove binding internals from public API

### LINKED ISSUE #9512: Make binding internals a private API
Our binding internals need a refactor, for a number of reasons (which I won't get into here), but we probably won't have chance to do this before 11.0, and so it will need to be done during the 11.0 timeframe.

The problem with that is that much of our binding internals are exposed as a public API meaning that we won't be able to refactor it without breaking our API promises. For that reason I think we'll need to make much of our binding internals a private API for 11.0, in particular the following classes:

- `InstancedBinding`
- `BindingExpression`
- `ExpressionObserver`
- `ExpressionNode` and derived classes
- The concrete implementations of `IPropertyAccessorPlugin`
- The concrete implementations of `IDataValidationPlugin`
- The concrete implementations of `IStreamPlugin`

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Makes various parts of our binding internals a private API as discussed in #9512.

## Breaking changes

The following types are now internal (removed from public API):

- `BindingExpression`
- `ExpressionObserver`
- `ExpressionNode` and derived classes
- The concrete implementations of `IPropertyAccessorPlugin`
- The concrete implementations of `IDataValidationPlugin`
- The concrete implementations of `IStreamPlugin`

The `InstancedBinding` API has been tweaked (not made internal as suggested in the issue due to feedback):

- Remove `Value` from the API, will always contain an `IObservable<object?>` from now
- The `IObservable
- Remove subject from the API, caller can try to cast the observable itself

## Fixed issues

Fixes #9512

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028759-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Data/InstancedBinding.cs [hunk 1]
```diff
@@ -106,7 +107,7 @@ internal InstancedBinding(object? value, BindingMode mode, BindingPriority prior
         {
             _ = observer ?? throw new ArgumentNullException(nameof(observer));
 
-            return new InstancedBinding(observer, BindingMode.OneWayToSource, priority);
+            return new InstancedBinding((IObservable<object?>)observer, BindingMode.OneWayToSource, priority);
```
Comment (maxkatz6) [inline src/Avalonia.Base/Data/InstancedBinding.cs:110]: Is it always a valid operation?
I can imagine having InstancedBinding with only IObserver implementation and without IObservable. And I suppose our binding system actually doesn't require IObservable in other places.

Comment (maxkatz6) [reply src/Avalonia.Base/Data/InstancedBinding.cs:110]: I wish it was possible to define API like:
```
public (IObservable | IObserver) Source { get; }
```
But that's not a TypeScript for better or worse.

**Review by maxkatz6 — APPROVED** [review]
LGTM in overall

File: src/Avalonia.Base/Data/InstancedBinding.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Base/Data/InstancedBinding.cs:110]: Yeah I did have that concern, but previously it used an `ISubject` so just an `IObserver` wouldn't have been possible anyway. Not too fond of this hack, but given that we're mainly keeping `InstancedBinding`  public for API compatibility it doesn't worry me that much,

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028859-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**